internal/encryption/keyprovider/openbao/provider.go

```diff
@@ -0,0 +1,66 @@
+package openbao
+
+import (
+	"context"
+
+	"github.com/opentofu/opentofu/internal/encryption/keyprovider"
+)
+
+type keyMeta struct {
+	Ciphertext []byte `json:"ciphertext"`
+}
+
+func (m keyMeta) isPresent() bool {
+	return len(m.Ciphertext) != 0
+}
+
+type keyProvider struct {
+	svc       service
+	keyName   string
+	keyLength DataKeyLength
+}
+
+func (p keyProvider) Provide(rawMeta keyprovider.KeyMeta) (keyprovider.Output, keyprovider.KeyMeta, error) {
+	if rawMeta == nil {
+		return keyprovider.Output{}, nil, &keyprovider.ErrInvalidMetadata{
+			Message: "bug: no metadata struct provided",
+		}
+	}
+
+	inMeta, ok := rawMeta.(*keyMeta)
+	if !ok {
+		return keyprovider.Output{}, nil, &keyprovider.ErrInvalidMetadata{
+			Message: "bug: invalid metadata struct type",
+		}
+	}
+
+	ctx := context.Background()
+
+	dataKey, err := p.svc.generateDataKey(ctx, p.keyName, p.keyLength.Bits())
+	if err != nil {
+		return keyprovider.Output{}, nil, &keyprovider.ErrKeyProviderFailure{
+			Message: "failed to generate OpenBao data key (check if the configuration valid and OpenBao server accessible)",
+			Cause:   err,
+		}
+	}
+
+	outMeta := &keyMeta{
+		Ciphertext: dataKey.Ciphertext,
+	}
+
+	out := keyprovider.Output{
+		EncryptionKey: dataKey.Plaintext,
+	}
+
+	if inMeta.isPresent() {
```

Comment:
Just making sure I understand the scheme correctly: every encryption, we'll use a fresh data key? I think this makes sense to me.

Reply:
Yes, this is exactly right. For every encryption, we use a fresh data key.